test/e2e/validation_test.go

```diff
@@ -96,3 +97,39 @@ func testUserData(t *testing.T) {
 	require.NoError(t, err, "could not find windows user data secret in required namespace")
 	assert.Contains(t, string(found.Data["userData"][:]), string(pubKeyBytes[:]), "expected user data not present in required namespace")
 }
+
+// testUserDataUpdate tests if userData reverts to previous value if updated
+func testUserDataUpdate(t *testing.T) {
+	secretInstance := &corev1.Secret{}
+	err := framework.Global.Client.Get(context.TODO(), kubeTypes.NamespacedName{Name: "windows-user-data", Namespace: "openshift-machine-api"}, secretInstance)
+	require.NoError(t, err, "could not find windows user data secret in required namespace")
+	validUserData := secretInstance.Data["userData"]
+
+	var tests = []struct {
+		name      string
+		operation string
+	}{
+		{"Update the userData secret with invalid data", "Update"},
+		{"Delete the userData secret", "Delete"},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.operation == "Update" {
+				updatedSecret := secretInstance.DeepCopy()
+				updatedSecret.Data["userData"] = []byte("invalid data")
+				err := framework.Global.Client.Update(context.TODO(), updatedSecret)
+				require.NoError(t, err, "could not update userData secret")
+				err = framework.Global.Client.Get(context.TODO(), kubeTypes.NamespacedName{Name: "windows-user-data", Namespace: "openshift-machine-api"}, secretInstance)
+				require.NoError(t, err, "could not find windows user data secret in required namespace")
+				require.Equal(t, validUserData, secretInstance.Data["userData"], "unable to revert userData Secret")
+			}
+			if tt.operation == "Delete" {
+				err := framework.Global.KubeClient.CoreV1().Secrets("openshift-machine-api").Delete(context.TODO(), "windows-user-data", metav1.DeleteOptions{})
+				require.NoError(t, err, "could not delete userData secret")
+				err = framework.Global.Client.Get(context.TODO(), kubeTypes.NamespacedName{Name: "windows-user-data", Namespace: "openshift-machine-api"}, secretInstance)
+				require.NoError(t, err, "could not find windows user data secret in required namespace")
```

Comment:
I would restructure this test block. [Table driven tests](https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go) are useful if you performing the same actions with different parameters and expect specific results. The way you are using it does not fall under this use case.

In order to make this test follow the table driven test pattern, you could change your `tests` struct definition to have: `{name, apiCall, expectedSecret}` where apiCall is a function pointer, and expectedSecret is the expected userData object.

If the above suggestion works thats cool, but if it doesn't theres no reason to try and force the table driven test pattern in a place it doesnt work. You can separate these tests into different test functions, so you'd have 
```
testUserdataDelete(t *testing.T) { ...}
testUserdataModify(t *testing.T) {...}
```
And then you could have a larger test that calls these two test functions as subtests to organize the test results:
```
testUserdataTamper(t *testing.T) {
t.Run("test that a user cannot delete the userdata secret", testUserdataDelete)
t.Run("test that a user cannot modify the userdata secret", testUserdataModify)
}
```

Reply:
I agree with most of what @sebsoto told except for if this should be a tabledriven test or not. To me, this is ok considering we are not writing units but e2es but we should have expectedResult for example if the test supposed to recreate a secret or not?